Add correct lat, lng to parsed GEOGRAPHY POINT [#1277]

jeremy-w opened the pull request:
Parsing now differentiates between parsing a geometry vs parsing a geography. Everything about the two structures is identical, except for their POINT structs, which have different valid ranges and, most significantly for us, different axis orders:

- Geography: y then x (Lat then Long)
- Geometry: x then y (no clever alternate names here, luckily)

However, in order to avoid breaking compatibility, this PR:

- Leaves the Geography x and y flipped
- But adds correct lat and lng fields in the parsed point object

Users of the library are recommended to migrate to relying on the lat and lng fields when working with geographies, in preference to the x and y fields.

jeremy-w commented:
This is rather light on the integration tests.

Should I also add an integration test that we can insert a point, parse it out, insert it per the parsed point info, and read it back again, with identical results? This would be easy to parameterize across geography vs geometry, since the WKT ordering of the axes is always x then y. Using values of x = 180 and y = 90, this would throw an error on insertion from the parsed struct if the axis ordering was flipped for the geography.

I could also add a test to verify that the geography point accessors .Lat and .Long agree with our parsed x and y from the overall geography column.

dhensby commented:
>Geography: y then x (Lat then Long)

Isn't the problem that it is (lng,lat) not (lat,lng)?

Regarding tests - I'm happy for you to cover off whatever you think is most appropriate :)

dhensby commented on test/common/tests.js:
```diff
@@ -971,7 +971,10 @@ module.exports = (sql, driver) => {
 
             if (isSQLServer2019OrNewer) {
-              expectedMessage = "String or binary data would be truncated in table 'master.dbo.tran_test', column 'data'. Truncated value: 'asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdfas'."
+              const config = JSON.parse(require('fs').readFileSync(join(__dirname, '../.mssql.json')))
```
Could this be `require(join(__dirname, '../.mssql.json'))`?

jeremy-w replied:
yup, worked a charm! and that's probably also caching it as a module and reading it just once, too. 👍

dhensby commented on lib/udt.js:
```diff
@@ -53,20 +53,35 @@ class Point {
-    point.y = buffer.readDoubleLE(buffer.position + 8)
-    buffer.position += 16
+  if (isGeometryPoint) {
+    // GEOGRAPHY POINT (s2.1.6): x then y.
```
Comment should be `GEOMETRY POINT`

jeremy-w replied:
🤦

dhensby commented on lib/udt.js:
```diff
@@ -53,20 +53,35 @@ class Point {
+    for (let i = 1; i <= count; i++) {
+      const point = new Point()
+      points.push(point)
+      point.y = buffer.readDoubleLE(buffer.position)
```
I'm wondering if for geography points it's worth either naming them lat/lng or adding a duplicate props of lat/lng. What do you think?

jeremy-w replied:
that's an interesting idea. and it avoids a lot of confusion.

as a bonus, we could leave the x/y flipped as before, and include correct lat/lng alongside, and then i believe it would no longer be a breaking change?

jeremy-w replied:
(and then possibly also do a follow-up breaking change that updates or eliminates the x/y in favor of lat/lng, but that would enable the fix landing in things that can't take a breaking change just now.)

dhensby replied:
great idea

jeremy-w replied:
ok, i've addressed the other issues. i will need to do a bit more work for this other direction, but i like it. :)

dhensby commented on test/common/tests.js:
```diff
@@ -3,6 +3,7 @@
 const assert = require('assert')
 const stream = require('stream')
 const { join } = require('path')
+const util = require('util')
```
can we pull just `format` out if that's all we are using (`const { format } = require('util');`)

dhensby commented on test/msnodesqlv8/msnodesqlv8.js:
```diff
@@ -178,7 +178,7 @@ describe('msnodesqlv8', function () {
   describe('msnodesqlv8 connection errors', function () {
     it('login failed', done => TESTS['login failed'](done, /Login failed for user '(.*)'\./))
-    it.skip('timeout (not supported by msnodesqlv8)', done => TESTS.timeout(done, /1000ms/))
+    it.skip('timeout (not supported by msnodesqlv8)', done => TESTS.timeout.call(this, done, /1000ms/))
```
I think this will need to be wrapped in a non-arrow anon function as with the tedious.js file

dhensby replied:
Pretty sure this still needs to be `it('timeout', function (done) { TESTS.timeout.call(this, done, /Failed to connect to 10.0.0.1:1433 in 1000ms/) })`

dhensby commented on test/tedious/tedious.js:
```diff
@@ -229,7 +229,8 @@ describe('tedious', () => {
     it('login failed', done => TESTS['login failed'](done, /Login failed for user '(.*)'/))
-    it('timeout', done => TESTS.timeout(done, /Failed to connect to 10.0.0.1:1433 in 1000ms/))
+    // call(this) to enable the test to skip itself.
+    it('timeout', function (done) { TESTS.timeout.call(this, done, /Failed to connect to 10.0.0.1:1433 in 1000ms/) })
```
Does this work, I didn't think you could bind new `this` context to arrow functions...

jeremy-w replied:
this works because TESTS.timeout is not an arrow function, but a method definition: `{ timeout(args) { body } }` rather than `{ timeout: (args) => { body } }`. see output in the node console of:

```js
t = { meth() { console.log(this) }, arr: () => { console.log(this) } }
t.meth.call('custom this')
t.arr.call('custom this')
```

arr has the globalThis anyway, but meth has 'custom this' as its this.

dhensby reviewed (changes requested):
This looks really good, and thanks for cleaning up some of the quirks with the test suite, it could do with a good going-over to sort some of its quirks.

Just a few minor comments for you at this point :)

jeremy-w commented:
@dhensby I believe I've addressed all the comments. I've updated the PR description to describe the backwards-compatible approach we decided on with adding correct lat/lng alongside unchanged (still flipped) x/y. And I extended the UDT integration test to confirm our notion of lat and lng agree with SQL Server's, as a final sanity check.

dhensby reviewed (approved):
Great! Thanks so much for this work, really excellent contribution.

Only other thing I can think is the changelog and readme updates if there are any needed?

jeremy-w commented:
@dhensby i've updated the README.

how do you handle adding changelog entries? in the past, i've accumulated changes in a Next Version and then updated the heading to record the version and release date as part of making a release.

dhensby commented:
I tried to push a commit for the changelog, but I don't have permission to push to your branch.

I'd just add this to the top of the changelog:

```
v7.2.0 (2021-??-??)
-------------------
[new] Update Geography field parsing to provide lat/lng props from Geography Point ((#1282)[https://github.com/tediousjs/node-mssql/pull/1282])
```

jeremy-w commented:
@dhensby I've added that text to the changelog.

I have "Allow edits by maintainers" checked, and that seems to include anyone who can push to the source of the fork, so I don't know why it didn't let you push. 🤔

dhensby commented:
Great work! I'll look to tag a minor release later today

dhensby commented:
Released in [v7.2.0](https://github.com/tediousjs/node-mssql/releases/tag/v7.2.0)
